pkg/cosigner/image_signer.go

```diff
@@ -0,0 +1,91 @@
+package cosigner
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"path/filepath"
+	"regexp"
+
+	"github.com/BurntSushi/toml"
+	"github.com/buildpacks/lifecycle"
+	"github.com/sigstore/cosign/cmd/cosign/cli"
+)
+
+type ImageSigner struct {
+	Logger *log.Logger
+}
+
+var (
+	cliSignCmd     = cli.SignCmd
```

Comment:
Asking for assistance and insight:

pkg/cosigner/image_signer.go: For testing purposes, we used cliSignCmd = cli.SignCmd so that we can override this function during testing to validate parameter input while ignoring its functionality. Is this an appropriate approach or is there another best practice? We didn't see or find any particular mocking/mock generating libraries being included or other examples in the code when we first looked.

Reply:
The thing I'm trying to wrap my head around right now is how would I use dependency injection if the function I am calling is not from an object. It is a public function, SignCmd, found [here](https://github.com/sigstore/cosign/blob/main/cmd/cosign/cli/sign.go)
In the example you provided, the factory object is set during instantiation so that it can be any object that matches the interface. However, the function in question for the above issue is a public function not attached to an object. I'm relatively newer to GoLang so please let me know if this understanding is wrong and that dependency injection can still work for this.

This function is what we wanted functionally, and I didn't want to re-implement it or copy/paste it into the code. 

Regarding your second point, I'll continue investigating

Reply:
In golang you can define a type which is itself a function.

That would look something like this: 
```go
type SignFunc func(ctx context.Context, ko cli.KeyOpts, annotations map[string]interface{}, imageRef string, certPath string, upload bool, payloadPath string, force bool, recursive bool) error

type ImageSigner struct {
	Logger *log.Logger
	signFunc SignFunc
}

func NewImageSigner(logger *log.Logger, signFunc SignFunc) *ImageSigner {
	return &ImageSigner{
		Logger: logger,
		signFunc: signFunc,
	}
}
```


Then the cosign func can be dependency injected like this in main:

```go
imageSigner := NewImageSigner(logger, cli.SignCmd)
```

Reply:
At this point the image signer is doing very little (which is a good thing) but, a test with a mocked out cosign is not giving us very much confidence that image signer is actually being invoked and used correctly. So, I would like to explore actually invoking the cosign library in the test. We test registry interactions in [client_test](https://github.com/pivotal/kpack/blob/main/pkg/registry/client_test.go) with a test http server and I imagine a cosign variation might look very similar. Happy to jump on a zoom to work on it together!

Reply:
It looks like sigstore/cosign has an [e2e_test.go](https://github.com/sigstore/cosign/blob/main/test/e2e_test.go) which runs a fake registry locally, pushes an image, then runs the cosign signCmd and verify on it. Would extracting the necessary stuff from this and re-implementing it in our testing be sufficient to validating that cosign's signCmd works?

Reply:
Going to make a commit first where I validate the real SignCmd once following the e2e_test.go sample. 
A second commit, which can be reverted if we see fit, where I update all the existing signCmd mocks to also run the real signCmd after validating the arguments. Afterwards, doing an additional validation that the expected items were pushed to the registry.